Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -6,102 +6,167 @@
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithRequest, in: subjectClass)
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithURLCompletion, in: subjectClass)
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithRequestCompletion, in: subjectClass)
-    }
+    enum DataTask_URL_Completion {
+        static let selector = #selector(URLSession.dataTask(with:completionHandler:) as (URLSession) -> (URL, @escaping CompletionHandler) -> URLSessionDataTask)
+        typealias TypedIMP = @convention(c) (URLSession, Selector, URL, @escaping CompletionHandler) -> URLSessionDataTask
+        typealias TypedBlockIMP = @convention(block) (URLSession, URL, @escaping CompletionHandler) -> URLSessionDataTask
 
-    static func swizzleDataTaskWithURL() throws {
-        // typealiases cannot be generic as C/block conventions don't support generics
-        // note that _Block doesn't have Selector parameter
-        typealias TypedIMP = TypedIMPs.DataTaskWithURL
-        typealias TypedBlockIMP = TypedBlocks.DataTaskWithURL
-        let sel = Selectors.DataTaskWithURL
+        static func swizzle(interceptor: @escaping RequestInterceptor) throws {
+            try swizzler.swizzle(selector: selector, in: subjectClass, with: newIMP(using: interceptor))
+        }
 
-        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+        static func unswizzle() throws {
+            try swizzler.unswizzle(selector: selector, in: subjectClass)
+        }
 
-        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURL -> URLSessionDataTask in
-            return typedOriginalImp(impSelf, impSelector, impURL)
+        private static func newIMP(using interceptor: @escaping RequestInterceptor) throws -> IMP {
+            let sel = selector
+            let redirectedSel = DataTask_Request_Completion.selector
+            let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+            let typedRedirectedImp: DataTask_Request_Completion.TypedIMP
+            typedRedirectedImp = try swizzler.originalImplementation(of: redirectedSel, in: subjectClass)
+
+            let newImpBlock: TypedBlockIMP = { impSelf, impURL, impCompletion -> URLSessionDataTask in
+                if let interceptionResult = interceptor(impSelf.urlRequest(with: impURL)) {
+                    weak var blockTask: URLSessionDataTask? = nil
+                    let modifiedCompletion: CompletionHandler = { origData, origResponse, origError in
+                        impCompletion(origData, origResponse, origError)
+                        blockTask?.payload?(.completed)
+                    }
+                    let task = typedRedirectedImp(impSelf, sel, interceptionResult.request, modifiedCompletion)
+                    // TODO: RUMM-452 report error?
+                    try? Resume.swizzle(in: task)
```

Comment:
I think the error handling will be easier to reason about if we centralize the swizzling code in a single place. Now, the swizzling is done in a deep-search way:
```swift
swizzleURLSession()
   → DataTask_URL_Completion.swizzle()
         → Resume.swizzleIfNotSwizzled()
   → DataTask_Request_Completion.swizzle()
         → Resume.swizzleIfNotSwizzled()
```

If we cut off the nesting and do it in a flat way:
```swift
swizzleURLSession()
   → let swizzleOnce: (URLSessionTask) -> Void = { task in URLSessionTaskSwizzler(task).swizzle() }
   → DataTask_URL_Completion.swizzle(: ..., followWith: { task in swizzleOnce(task) })
   → DataTask_Request_Completion.swizzle(: ..., followWith: { task in swizzleOnce(task) })
```
this will be IMO much easier to follow. Other benefits:
* the `swizzleIfNotSwizzled()` method will be no longer needed (we can easily make `swizzleOnce` closure being called only once in the local scope);
* the code will be more decoupled, into: `URLSessionSwizzler` + `URLSessionTaskSwizzler`;
* `URLSessionTask` swizzling and its errors can be tested separately.

Reply:
i considered that but it didn't turn out to be as good as i imagine

1. currently, the new implementation of `dataTask...` method is in the same place as a whole. we don't need to look at different places to understand what it does. IMO this is easier to read.
2. swizzled` resume` implementation relies on `task.payload` associated object, payload is set by `dataTask...` implementation; so they always have this implicit coupling

i think i didn't understand how this proposal would help with decoupling 🤔 can you explain a bit more please?

i would avoid `swizzleOnce` because there is no public API guarantee that the `dataTask` type will always be the same: `URLSession` has different configs and i suppose different configs may produce different data task types somehow

Reply:
> i think i didn't understand how this proposal would help with decoupling 🤔 can you explain a bit more please?

I was thinking that if both:
* `session.dataTaskWithURLCompletion()`
* and `session.dataTaskWithURLRequestCompletion()`

can result with either "swizzling OK" or "swizzling FAIL" and the `task.resume()` swizzling can result with "OK" or "FAIL" this builds up 8 control flows to test. By separating task swizzling from session swizzling, there would be only 2 scenarios in the prior and 4 in the latter. But turns out that my assumption is not true, as "session FAIL, resume OK" is not a possible control flow.

Let's keep it the way it is 👌.

> i would (NOT) avoid swizzleOnce because there is no public API guarantee that the dataTask type will always be the same: URLSession has different configs and i suppose different configs may produce different data task types somehow

I see, sounds fair enough 👍.

Reply:
i'm a bit confused, your quote my comment with adding `NOT` in the most important part :) but i'm assuming we are clear on this
otherwise, feel free to unresolve 👍